mmrotate/models/dense_heads/csl_rotated_retina_head.py

```diff
@@ -0,0 +1,564 @@
+        for cls_score, bbox_pred, angle_cls, anchors in zip(
+                cls_score_list, bbox_pred_list, angle_cls_list, mlvl_anchors):
+            assert cls_score.size()[-2:] == bbox_pred.size()[-2:]
+            cls_score = cls_score.permute(1, 2,
+                                          0).reshape(-1, self.cls_out_channels)
+            if self.use_sigmoid_cls:
+                scores = cls_score.sigmoid()
+            else:
+                scores = cls_score.softmax(-1)
+            bbox_pred = bbox_pred.permute(1, 2, 0).reshape(-1, 5)
+
+            angle_cls = angle_cls.permute(1, 2, 0).reshape(
+                -1, self.coding_len).sigmoid()
+            angle_pred = self.angle_coder.decode(angle_cls)
+
+            nms_pre = cfg.get('nms_pre', -1)
+            if scores.shape[0] > nms_pre > 0:
+                # Get maximum scores for foreground classes.
+                if self.use_sigmoid_cls:
+                    max_scores, _ = scores.max(dim=1)
+                else:
+                    # remind that we set FG labels to [0, num_class-1]
+                    # since mmdet v2.0
+                    # BG cat_id: num_class
+                    max_scores, _ = scores[:, :-1].max(dim=1)
+                _, topk_inds = max_scores.topk(nms_pre)
+                anchors = anchors[topk_inds, :]
+                bbox_pred = bbox_pred[topk_inds, :]
+                scores = scores[topk_inds, :]
+                angle_pred = angle_pred[topk_inds]
+            # csl decoder
+            if self.use_encoded_angle:
+                bbox_pred[..., -1] = angle_pred
+                bboxes = self.bbox_coder.decode(
+                    anchors, bbox_pred, max_shape=img_shape)
+            else:
+                bboxes = self.bbox_coder.decode(
+                    anchors, bbox_pred, max_shape=img_shape)
+                bboxes[..., -1] = angle_pred
+
+            mlvl_bboxes.append(bboxes)
+            mlvl_scores.append(scores)
+        mlvl_bboxes = torch.cat(mlvl_bboxes)
+        if rescale:
+            # angle should not be rescaled
+            mlvl_bboxes[:, :4] = mlvl_bboxes[:, :4] / mlvl_bboxes.new_tensor(
+                scale_factor)
+        mlvl_scores = torch.cat(mlvl_scores)
+        if self.use_sigmoid_cls:
+            # Add a dummy background class to the backend when using sigmoid
+            # remind that we set FG labels to [0, num_class-1] since mmdet v2.0
+            # BG cat_id: num_class
+            padding = mlvl_scores.new_zeros(mlvl_scores.shape[0], 1)
+            mlvl_scores = torch.cat([mlvl_scores, padding], dim=1)
+
+        if with_nms:
+            det_bboxes, det_labels = multiclass_nms_rotated(
+                mlvl_bboxes, mlvl_scores, cfg.score_thr, cfg.nms,
+                cfg.max_per_img)
+            return det_bboxes, det_labels
+        else:
+            return mlvl_bboxes, mlvl_scores
+
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds', 'angle_classes'))
+    def get_bboxes(self,
```

Comment:
`angle_classes` doesn't apear in inputs.

Reply:
It looks like a typo.